### PR TITLE
optimize-concurrent-downloads

### PR DESCRIPTION
**Pull request by oliviercommelarbre**
Hi Tatsuhiro, as anticipated here the proposed update to support dynamic tuning/management of the number of concurrent downloads as a function of the download bandwidth observed. Still a few bits to complete in the option handling in particular but i would need there your advice. 
thanks for looking at it

### CODE REVIEW

**Comment by tatsuhiro-t** [comment]
Awesome! Thank you.  Will review this PR in this weekend.  Please note that we are now in feature/documentation freeze period now (next release is February 15).  The actual merge will be done after the release.

**Comment by luokar** [comment]
I agree some sort of download number optimization is great, but I still don't understand why using
the formula N = A + B Log10(speed in Mbps) to determine N. Somethings, we might need to download  a large number of small file which will slow down current total speed, but decreasing N will just make things worse. 
I propose instead of using N = A + B Log10(speed in Mbps), we should compare the number of download and its effect on total speed or something like an AIMD algorithm.

**Comment by oliviercommelarbre** [comment]
to Tatsuhiro: Thanks! and no pb about missing the 15/02 release, it would get a chance in the next one

when you review could you please take a specific look at the option, i have proposed the following template for the final implementation (see in usage_text.h): 
--optimize-concurrent-downloads[=true|false|NUM:NUM]
but i've only implemented it as a simple boolean (true | false) for the time being (at present A and B are hardcoded). 
the NUM:NUM case would mean essentially 'true' while customizing the A and B parameters of the optimization function at the same time.

before completing this part, I wanted to get first your advice on it e.g. does it fit enough the standard of aria2c options? or would you recommend  setting A and B through a separate option? or may be you would have yet another recommendation.

thanks

**Comment by oliviercommelarbre** [comment]
To Luokar:

it is a good point and in fact initially i was more looking for something in the idea that you mention, like an adaptive/progressive algo that assesses/reacts on the effect on the overall download speed of adding/removing parallel downloads. 
however i found out that it would be extremely complex to get a good value somehow, here the rationale in a few words:

- assessing the benefits of adding a parallel download will be very complex. first it will depend on the initial latency to open the connection before you receive the first byte that will have a chance to influence the download speed in a tangible manner. e.g.  you would have to wait some time before you could conclude on whether it was a good thing or bad thing. This wait time will depend on the protocol you use and on the latency of the network you are on, which will make the real-time algo even more complex. also, waiting would imply that the algo will be less reactive at the beginning. e.g. if you have a 50mbps link, you aim at going straight to several 10s of parallel downloads to saturate your inbound bandwith and optimise the overall session. going step by step after a careful adaptive algorithm might result in getting to saturate the inbound bandwith only after several 10s of seconds or more. Also i found that it will be difficult to know whether the speed increase you could measure is effectively due to your additional parallel download(s), or just a result of your available bandwith augmenting for other reasons which brings additional complexity.

- second, i found out that the gain on the overall download speed of adding parallel downloads does not necessarily require a fine control. Practically what we want to achieve is 
(1) to use effectively the idle time between successive downloads (end a connection, start a new one) during which the inbound bandwith is not well used. having a parallel download at the time of these idles of the other connections ensures that the inbound bandwidth is always used to some extent.
(2) to slow down every single TCP connection so that the effect of its latency has less impact on the overall download throughput we get out of it.
(3) to avoid client congestion and server timeouts that go in pair when too many parallel downloads are set (the many parallel connections end up going so slowly by sharing the same inbound bandwidth that the corresponding servers aborts before the download completes)
My conclusion here was that the number of parallel downloads we should be using is more a function of the class of performance of the network we are using, rather than requiring a fine control loop. 
the reason for using the logarithm is exactly following this rationale. Adding a degree of magnitude to your inbound network bandwidth should result into adding parallel downloads nearly linearly to the difference in degrees of magnitude.

The other good thing is that the range of performance of the network can be sensed nearly immediately after a download starts. i.e. a second is enough to figure out roughly the right range, which is an asset to ensure the dynamics of the intended optimization.

Having said that you are right that the download of very small files will lead to reaching less easily the available bandwidth (because of the latency and the connection overhead), hence the number of parallel downloads should be sustained slightly higher than the actual need for big ones. Puting it slightly higher than needed will have i think very little negative impact on the download performance for big files, the important thing in my view is to get it roughly right (to achieve the goals 1, 2 and 3) and slightly more than stricly needed (to sustain a good performance in all cases). I've proposed very draft default values for A and B. in fact i have tested the algo on files of a few megabytes each (2 to 6 MB typically) and on 1 to 100Mbps networks typically. 
We could try to see whether we should expect a big dependency of these A&B settings wrt the size of the individual files, it would be in fact interesting to validate the approach and may be tune the default parameters.

File: src/RequestGroupMan.h [hunk 1]
```diff
@@ -195,6 +198,18 @@ class RequestGroupMan {
 
   bool removeReservedGroup(a2_gid_t gid);
 
+  bool getOptimizeSimultaneousDownloads() const
+  {
+    return optimizeSimultaneousDownloads_;
+  }
+
+  void setOptimizeSimultaneousDownloads(bool flag)
```
Comment (tatsuhiro-t) [inline src/RequestGroupMan.h]: It seems this is not used. If yes, please remove this.

File: src/RequestGroupMan.h [hunk 2]
```diff
@@ -195,6 +198,18 @@ class RequestGroupMan {
...
+  {
+    optimizeSimultaneousDownloads_=flag;
+  }
+
+  bool setOptimizeSimultaneousDownloads();
```
Comment (tatsuhiro-t) [inline src/RequestGroupMan.h]: Usually, we use setSomething to set parameter to private field.  Since this function has no parameters, perhaps, something like setupOptimizeSimultaneousDownloads would sound more like it.

File: src/RequestGroupMan.cc [hunk 3]
```diff
@@ -474,11 +484,43 @@ createInitialCommand(const std::shared_ptr<RequestGroup>& requestGroup,
...
+      // apply the rule
+      int speed=currentSpeed;
+      if((maxOverallDownloadSpeedLimit_>0) && (speed>maxOverallDownloadSpeedLimit_)) {
+        speed=maxOverallDownloadSpeedLimit_;
+      }
+      maxSimultaneousDownloads=ceil(
+        optimizeSimultaneousDownloadsCoeffA_
+        +optimizeSimultaneousDownloadsCoeffB_ * log10(speed*8./1000000.)
+      );
+      if((maxSimultaneousDownloads_>0)&&(maxSimultaneousDownloads>maxSimultaneousDownloads_)) {
```
Comment (tatsuhiro-t) [inline src/RequestGroupMan.cc]: maxSimultaneousDownloads_ is always more than or equals to 1, so checking >0 is not necessary.

File: src/RequestGroupMan.cc [hunk 4]
```diff
@@ -474,11 +484,43 @@ createInitialCommand(const std::shared_ptr<RequestGroup>& requestGroup,
...
+        maxSimultaneousDownloads=maxSimultaneousDownloads_;
+      }
+      if(maxSimultaneousDownloads<1) {
```
Comment (tatsuhiro-t) [inline src/RequestGroupMan.cc]: We can first check this condition before the above if statement to make the intention clear.  Something like:
```cpp
if (maxSimultaneousDownloads < 1) {
  maxSimultaneousDownloads = 1;
} else if (maxSimultaneousDownloads > maxSimultaneousDownloads_) {
  maxSimultaneousDownlaods = maxSimultaneousDownloads_;
}
```
or just go in one-liner:
```cpp
maxSimultaneousDownloads = std::min(std::max(1, maxSimultaneousDownloads), maxSimultaneousDownloads_);
```

File: src/RequestGroupMan.cc [hunk 5]
```diff
@@ -474,11 +484,43 @@ createInitialCommand(const std::shared_ptr<RequestGroup>& requestGroup,
...
+        maxSimultaneousDownloads=1;
+      }
+      A2_LOG_DEBUG
+        (fmt("Number of concurrent downloads adapted to speed (%d B/s) with max %d"
+             " (%lu currently active)",
+             speed, maxSimultaneousDownloads, numActive_));
```
Comment (tatsuhiro-t) [inline src/RequestGroupMan.cc]: Perhaps, it would be nice to use `util::abbrevSize(speed).c_str()`, and replace `speed (%d B/s)` with `speed (%s B/s)` in case where speed gets faster.

**Comment by tatsuhiro-t** [comment]
Overall, this is very good PR.  Thank you.  I like this idea.  @luokar has the point for the tiny file downloads, but this feature is optional, and optimizing relatively large downloads would be also plus.  Let's see how this feature does not go well with smaller downloads, and get some idea to improve it if it is ever required.

For options, it is OK as long as the usage is documented there.

File: src/RequestGroupMan.cc [hunk 6]
```diff
@@ -105,6 +105,9 @@ RequestGroupMan::RequestGroupMan(
     int maxSimultaneousDownloads, const Option* option)
     : maxSimultaneousDownloads_(maxSimultaneousDownloads),
       numActive_(0),
+      optimizeSimultaneousDownloads_(false),
+      optimizeSimultaneousDownloadsCoeffA_(5.),
+      optimizeSimultaneousDownloadsCoeffB_(25.),
```
Comment (tatsuhiro-t) [inline src/RequestGroupMan.cc]: The above 3 lines should be put before `numActive_(0),`.   Compiler shows warnings like `field 'numActive_' will be initialized after field 'optimizeSimultaneousDownloads_' [-Werror,-Wreorder]`

File: src/RequestGroupMan.h (hunk 2 above)
Comment (oliviercommelarbre) [inline src/RequestGroupMan.h]: agreed with that, will do

File: src/RequestGroupMan.cc (hunk 3 above)
Comment (oliviercommelarbre) [inline src/RequestGroupMan.cc]: you are right. in fact i remember this was temporary... what i intended initially was to check whether --max-concurrent-downloads had been explicitly defined, but i did not find the right way to do it and left this in. 
The idea would be that if max-concurrent-downloads is not explicitly set, then there is no absolute maximum for the optimization. I would find it useful, please tell me your opinion and i would  complete. I found later on that every option implements a defined() method and i guess it returns false if not explicitly defined.

File: src/RequestGroupMan.cc (hunk 4 above)
Comment (oliviercommelarbre) [inline src/RequestGroupMan.cc]: good point, will update

File: src/RequestGroupMan.cc (hunk 5 above)
Comment (oliviercommelarbre) [inline src/RequestGroupMan.cc]: thanks for the tip, much nicer, will do

File: src/RequestGroupMan.cc (hunk 6 above)
Comment (oliviercommelarbre) [inline src/RequestGroupMan.cc]: ok, strange i don't remember my compiler giving me the warning

File: src/RequestGroupMan.cc (hunk 3 above)
Comment (tatsuhiro-t) [inline src/RequestGroupMan.cc]: The initial value of PREF_MAX_CONCURRENT_DOWNLOADS is 5 (https://github.com/tatsuhiro-t/aria2/blob/2a0b4751f4bccb198899be52b2c45249a79de5ab/src/OptionHandlerFactory.cc#L414), and its acceptable lowest value is 1.  And its value is passed to RequestGroupMan constructor.  So it has always be explicitly defined.  The maximum value gets quite big, and users can specify big value if they want.

File: src/RequestGroupMan.h (hunk 1 above)
Comment (oliviercommelarbre) [inline src/RequestGroupMan.h]: true it is not used (only the other one, now renamed with 'setup' as per your suggestion is used), i had just put it for completeness of set/get, but i will remove.

File: src/RequestGroupMan.cc (hunk 3 above)
Comment (oliviercommelarbre) [inline src/RequestGroupMan.cc]: ok!

**Comment by oliviercommelarbre** [comment]
Hi Tatsuhiro, i'm glad you like it and thks for giving credit. In the meantime I've been test bedding the case highlighted by @luokar, and doing that i found out that the speed measurements i get are sometimes inconsistent, as they vary unexpectedly jumping from hundreds of KiB/s to suddenly tens or few KiB/s within the same second of invoking the calculateSpeed method. This is in fact not specifically linked to the case of very small files although the case happens more often in that case since we get more often in the loop due to downloads finishing quickly.
I've debugged a bit and i think i found out the source of the problem. The SpeedCalc class which is in charge of computing the speed uses a timetag for the speed measurement updates (now = global::wallclock()) that is refreshed only once in between every command execution (within DownloadEngine::run). I need to study a little more to understand a good way round to avoid the inconsistency i observed. I'll come back to you...if you see an easy way round please don't hesitate to comment

File: src/FillRequestGroupCommand.cc [hunk 7]
```diff
@@ -80,6 +80,17 @@ bool FillRequestGroupCommand::execute()
     }
   }
   e_->addRoutineCommand(std::unique_ptr<Command>(this));
+
+  // let's make sure we come back here every second or so
+  // if we use the optimize-concurrent-download option
+  if(rgman->getOptimizeSimultaneousDownloads()) {
+    const Timer now;
```
Comment (tatsuhiro-t) [inline src/FillRequestGroupCommand.cc]: If possible, please use global::wallclock(), which saves many syscalls to get time stamp.

Comment (oliviercommelarbre) [inline src/FillRequestGroupCommand.cc]: done

**Comment by oliviercommelarbre** [comment]
Hi Tatsuhiro, i think i have completed as per the thread sofar + the option processing + an improvement to the algo to cover better for the case of small file downloads. I'm quite happy sofar with the results of my tests. i let you take a look.

**Comment by oliviercommelarbre** [comment]
small addendum: i'm not very familiar with github.  i've commited in my initial branch (issued from 1.19.3) which is no more the aria2c master since the last merge of the 15/02. please let me know if that's ok or if i should do it differently, thks

File: src/OptionHandlerImpl.cc [hunk 8]
```diff
@@ -586,6 +586,61 @@ std::string PrioritizePieceOptionHandler::createPossibleValuesString() const
...
+    : AbstractOptionHandler(pref, description, defaultValue,
+                            OptionHandler::OPT_ARG, shortName)
+{
+}
+
+void OptimizeConcurrentDownloadsOptionHandler::parseArg(Option& option,
+                                     const std::string& optarg) const
+{
+  if (optarg == "true" || ((argType_ == OptionHandler::OPT_ARG ||
+                            argType_ == OptionHandler::NO_ARG) &&
```
Comment (tatsuhiro-t) [inline src/OptionHandlerImpl.cc]: We set argType_ to OptionHandler::OPT_ARG in constructor L592, so we don't have to check that value here.

File: src/OptionHandlerImpl.cc [hunk 9]
```diff
@@ -586,6 +586,61 @@ std::string PrioritizePieceOptionHandler::createPossibleValuesString() const
...
+    PrefPtr pref, const char* description, const std::string& defaultValue, char shortName)
+    : AbstractOptionHandler(pref, description, defaultValue,
+                            OptionHandler::OPT_ARG, shortName)
+{
+}
+
+void OptimizeConcurrentDownloadsOptionHandler::parseArg(Option& option,
+                                     const std::string& optarg) const
+{
+  if (optarg == "true" || ((argType_ == OptionHandler::OPT_ARG ||
```
Comment (tatsuhiro-t) [inline src/OptionHandlerImpl.cc]: Please use A2_V_TRUE for "true" string literal, and A2_V_FALSE for "false".

File: src/OptionHandlerImpl.cc [hunk 10]
```diff
@@ -586,6 +586,61 @@ std::string PrioritizePieceOptionHandler::createPossibleValuesString() const
...
+    std::string *sptr = &coeff_a;
+    for(;;) {
+      try {
+        double dbl=std::stod (*sptr);
+      } catch(std::invalid_argument &ex) {
+        throw DL_ABORT_EX(fmt("Bad number '%s'", sptr->c_str()));
+      }
+      option.put(pref,*sptr);
+
+      if(pref==PREF_OPTIMIZE_CONCURRENT_DOWNLOADS_COEFFB) break;
```
Comment (tatsuhiro-t) [inline src/OptionHandlerImpl.cc]: We usually use {} even with single line.

File: src/usage_text.h [hunk 11]
```diff
@@ -242,7 +242,21 @@
 #define TEXT_MAX_CONCURRENT_DOWNLOADS                                   \
   _(" -j, --max-concurrent-downloads=N Set maximum number of parallel downloads for\n" \
     "                              every static (HTTP/FTP) URL, torrent and metalink.\n" \
-    "                              See also --split option.")
+    "                              See also --split and --optimize-concurrent-downloads options.")
+#define TEXT_OPTIMIZE_CONCURRENT_DOWNLOADS\
+  _(" --optimize-concurrent-downloads[=true|false|NUM:NUM] Optimizes the number of\n" \
```
Comment (tatsuhiro-t) [inline src/usage_text.h]: We are referring first NUM as A, and second NUM as B.  I think it would be clearer to say `A:B` rather than `NUM:NUM`.

File: src/OptionHandlerImpl.cc [hunk 12]
```diff
@@ -586,6 +586,61 @@ std::string PrioritizePieceOptionHandler::createPossibleValuesString() const
...
+      pref=PREF_OPTIMIZE_CONCURRENT_DOWNLOADS_COEFFB;
+      sptr = &coeff_b;
+    }
+    option.put(pref_, A2_V_TRUE);
+  }
+}
+
+std::string OptimizeConcurrentDownloadsOptionHandler::createPossibleValuesString() const
+{
+  return "true, false, NUM:NUM";
```
Comment (tatsuhiro-t) [inline src/OptionHandlerImpl.cc]: We are referring first NUM as A, and second NUM as B in help message.  I think it would be clearer to say `A:B` rather than `NUM:NUM`.

File: src/RequestGroupMan.cc [hunk 13]
```diff
@@ -1005,4 +1026,58 @@ void RequestGroupMan::decreaseNumActive()
...
+  // gauge the current speed
+  int currentSpeed=getNetStat().calculateDownloadSpeed();
+
+  // estimate the actual bandwidth consumed including connection overheads, assuming 4KiB of overhead for each connection initialisation
+  int bandwidth=currentSpeed;
+  double speedUndervalueRatio(0.);
+  uint64_t dl(getNetStat().getSessionDownloadLength());
+  if(dl>0) {
+    speedUndervalueRatio=4096.*(numActive_+numStoppedTotal_)/dl;
+    bandwidth*=(1+speedUndervalueRatio);
```
Comment (tatsuhiro-t) [inline src/RequestGroupMan.cc]: Could you explain the theory behind these calculations?

File: src/OptionHandlerImpl.cc (hunk 8 above)
Comment (oliviercommelarbre) [inline src/OptionHandlerImpl.cc]: you are right (i just copied the case of the Boolean handler and left that in), will correct

File: src/usage_text.h (hunk 11 above)
Comment (oliviercommelarbre) [inline src/usage_text.h]: ok will do, i had a doubt myself on the convention to take

File: src/SpeedCalc.cc [hunk 14]
```diff
@@ -84,6 +84,29 @@ int SpeedCalc::calculateSpeed()
...
+
+  int64_t bytesCount(0);
+  auto it=timeSlots_.rbegin();
+  while (it != timeSlots_.rend()) {
+    if (it->first.difference(now) > seconds*1_s) {
+      break;
+    }
+    bytesCount += (*it++).second;
+  }
+  if (it == timeSlots_.rbegin()) return 0;
```
Comment (tatsuhiro-t) [inline src/SpeedCalc.cc]: I think we should return the sum of received bytes including first slot.  If all slots are within given seconds, all slots should be considered.

**Comment by tatsuhiro-t** [comment]
Thank you!  I commented inline.  Could you check them out?

To make the branch up to date, you can merge aria2 master into your branch.
I think you can do `git pull https://github.com/tatsuhiro-t/aria2.git` on your local branch.

**Comment by tatsuhiro-t** [comment]
Looks like your branch can be merged cleanly.

File: src/OptionHandlerImpl.cc (hunk 10 above)
Comment (oliviercommelarbre) [inline src/OptionHandlerImpl.cc]: ok, will follow the convention

File: src/RequestGroupMan.cc (hunk 13 above)
Comment (oliviercommelarbre) [inline src/RequestGroupMan.cc]: you are quite wise in raising that question ;-), it made me do some additional tests and these tests made me conclude that i don't have a good answer!
If you are curious, please read along. otherwise you can ignore since my new proposal would be to remove this part. i found out since then that the concept i was trying to model is actually not necessary to meet the objective it was supposed to contribute to...

the original objective was to take into account the 'overhead'  imposed by each of the file transfers on the speed measured, noting that the speed reached by aria2 in the case of small files is much lower compared to that of large files, even in the case of a high number of parallel downloads, which in turns makes the algo conclude on a too small number of parallel dowloads at the next iteration (the case highlighted by luokar).

the rationale for this calculation was hence to compute the actual bandwidth utilized instead of just measuring the effective download speed. As a matter of fact, the 'speed' computed by aria2 in SpeedCalc is measuring the actual number of 'useful' file bytes received by the client in a second of time after the first 'useful' byte has been received for every file download active, which is not necessarily linear with the actual bandwidth utilized if one is to consider the initial traffic necessary to initialise each transfer (e.g.ftp or http connection initialisation overhead) before the first 'useful' byte is received.
 
The speedUndervalueRatio was here to measure this overhead on the principle that every file transfer, regardless of the size of the file, would impose a fixed traffic on the link of N initialisation bytes before the first useful byte is received and accounted for in SpeedCalc.
Along this principle:
 speedUndervalueRatio = bandwidth/speed-1 
= (bytesWindow_ + N*(number of downloads active in the speed elapsed period))/bytesWindow_ -1 
= N*(number of downloads active in the speed elapsed period)/bytesWindow_

in turn, and on the assumption that the overhead would have limited variability in a session (sizes of files equally distributed along the session time), the overhead could be approximated as:
speedUndervalueRatio ~ N*(total number of downloads since the start of the session)/(total number of bytes received since the start of the session)

after my empirical tests, i had concluded for a value for N in the order of 4 KiB so that the bandwidth computed on 1KiB parallel ftp file transfers could be comparable to that of 1MiB parallel ftp file transfers.

---------------------
Now, back to reality after my additional tests:
- i found out that in case of http, the overhead would need to be much less than the 4KiB which i had benchmarked on ftp transfers. 
- trying to reach a higher number of parallel downloads for small files as compared to the case of large files is not necessarily a good thing since it can tire up the server(s) which have to process too many requests too frequently (the frequency of requests in a scenario with many small files is larger since they are downloaded quickly)
- finally and more importantly, the main unwanted scenario in the case of small files is already taken care of by the Timer of 5 seconds that waits for the parallel downloads to ramp-up the speed while ignoring possible drops of speed while the steady state is being reached. Also, monitoring that the speed keeps up well before decreasing number of parallel downloads is warrant that the steady state is maintained even for small file downloads.

in conclusion, i propose to roll back to using the speed directly which gives equally good results in both ftp and http cases.

File: src/SpeedCalc.cc (hunk 14 above)
Comment (oliviercommelarbre) [inline src/SpeedCalc.cc]: i'm not sure to understand. 
all slots that have a timetag in the range are considered.
the 'if' on rbegin is there in case the timeSlots_ list is empty or if the while loop exits immediately (it is actually unlikely but one never knows). Note that the while loop exits normally with one position ahead on the iterator, hence if the iterator has not moved from its first rbegin position, then we are in this case and we have no slot to compute the speed upon.
please comment back if i missed something, thanks

Comment (tatsuhiro-t) [inline src/SpeedCalc.cc]: You are completely right.  I was wrong, probably was confused about rbegin() thing.  Sorry for that.  This hunk is all good.

**Comment by nmaier** [comment]
>Files changed 4,422

Something went wrong here...
@oliviercommelarbre, seems you committed your build directory?

**Comment by oliviercommelarbre** [comment]
yes it seems so... and unexpectedly so. It seems my build-release directory under /aria2 went all in for whatever reason... i'll try to do a new commit that removes it.

**Comment by oliviercommelarbre** [comment]
@nmaier it should be ok now, sorry about that

**Comment by nmaier** [comment]
`make clang-format` is probably a good idea.
In general, white space is king; `if (abc == def) for (ghi; ghi < jkl; jkl) mno = pqr;`

**Comment by tatsuhiro-t** [comment]
`make clang-format` is good thing.  If clang-format is not available, then we can do it for you.
Other than that, PR looks good and ready for merge.
I'd like to avoid too many diff introduced in the build directory (which was fixed). Can you rebase and squash the commits into one?  If you are OK, we can do it on my side before merge.  It does not change the authorship of the commit, but only the first commit comment is retained.

**Comment by oliviercommelarbre** [comment]
Hi Tatsuhiro and nmaier, sorry for the silence, i'm in a middle of a house restructuration so my week ends are a bit busy... i've no problem about merging all commits into one and i understand it can make it cleaner after the spurious commit of the build directory...
For revieweing the code according to the last comment of nmaier no pb either, i think i can find a bit of time tomorrow if that's ok.

**Comment by tatsuhiro-t** [comment]
Thank you.  OK, then we'll rebase the commits when everything is OK.
Unfortunately, we entered feature/documentation freeze period (again), so merge will be done after the next release date, March 15.

**Comment by oliviercommelarbre** [comment]
Hi Tatsuhiro and Nils, i've just commited the syntax corrections, so it should be ready for the rebase. i wonder if we can make it on time for the merge before 15/03? otherwise no pb, we should be ready for the next period. cheers to both

**Comment by tatsuhiro-t** [comment]
Thank you.  Sorry, we have now in documentation/feature freeze period, which started March 5.  Since we have coordinated manual translation, we cannot add extra feature in this stage.

**Comment by tatsuhiro-t** [comment]
Merged as 9e05371fb6a70d7d5e9d49db31b879e79c1b56e7
Thank you very much!

**Comment by oliviercommelarbre** [comment]
Thank You Tatsuhiro, it's been a real pleasure, we'll be eager to do more, thanks also for managing this so nicely and easily
